### DISCUSSION TITLE
Is a breaking change for the serialized signature fine?

### DISCUSSION THREAD

**Discussion by CosmicHorrorDev** [dropbox/fast_rsync#17] [Ideas]
So I think it will flow better for me to explain the why before the what.

For when diffing identical or like files the computations for `Signature::calculate` are very similar to what is required for `diff`, yet there is a large difference in speed between both operations because `Signature::calculate` leverages SIMD instructions while `diff` does not. From thinking over things there are two different ways that things can be changed so that `diff` can also use SIMD instructions for a very nice speedup on same/like files

## Chunking input to MD4

_Note: this is the one that would be a breaking change for serialized signatures, but I don't think it breaks any other part of the public API_

The idea here is that we can change how we hash a block for MD4 (it won't actually be MD4 with this change of course) where instead of hashing the whole block you:

- Split the block into `crate::md4::simd::MAX_LANES` chunks
- Hash all the chunks utilizing the SIMD implementation
- Hash all the output hashes of the last step.

Doing this makes it so that even hashing a single block can still utilize SIMD instructions. It would require some special handling regarding if block-sizes are used that aren't a multiple of 8 or if very very small block-sizes are used

## Storing a buffer when CRC matches

_Note: unlike the above this won't be a breaking change_

The implementation gets a bit more complicated here (I had a branch that seemed to work most of the time, but still had some bugs. Sadly I accidentally deleted the branch before I fixed all the bugs). The idea here is that when `diff`ing specifically we have a buffer of upto `crate::md4::simd::MAX_LANES` blocks. Instead of scanning through the file, checking for CRC matches, on CRC match checking for MD4 match we would:

- Check for CRC match
- On match store the CRC value and index in the buffer
  - If the buffer is full then use the SIMD implementation to hash the blocks. Comparing the MD4 hashes and moving blocks accordingly
- When there isn't a match then use the SIMD implementation on any blocks in the buffer still

The way I was implementing things before went essentially as follows

```
outer-loop loop {
    move-by-block loop {
        get CRC value
        if crc match {
            append match info to buffer
        }

        if not a crc match or buffer is full {
            md4 hash the blocks and compare the hashes
            if not a crc match or one of the md4 hashes didnt match {
                break move-by-block
            }
        }

        move index by block
    }

    update the crc value and index accordingly

    move-by-byte {
        move index by byte
        get CRC value
        if crc match {
            break move-by-byte
        }
}
```

I elided going over handling hitting the end of file along with other cases to keep things simple, but there is a lot of special case handling in different places as I'm sure you can imagine. Splitting things up into two inner loops mitigated a lot of performance impact when there are few matching chunks between the two inputs

---

I'd like to pursue one of these two routes, I just wanted some input before I got into things again. This also means that we would get to have another win for this situation covered in the documentation :eyes:

> When comparing similar files, `fast_rsync` is mostly bound by the speed of single-block MD4 hashing, so it is not much faster than `librsync`.

**Comment by goffrie** [comment]
> Chunking input to MD4

I don't love this idea - part of the point of this library is that it's compatible with librsync (at least in some limited way). Also, I think if you wanted to change the hash algorithm, switching to something like BLAKE3 would make more sense; it's a much more modern hash design and it's also able to make use of SIMD parallelism within a single hash input.

> Storing a buffer when CRC matches

So I thought about this too (or something like it, anyhow), but it wasn't clear to me that this would make things faster "overall". The core idea if I understand correctly is to "speculatively" assume a MD4 match and continue CRC matching until we accumulate enough blocks to hash the previous matches in a batch, then rewind any that failed. I worry that this could lead to a lot of extra work being done if there are frequent false positives (the CRC is *extremely* weak, remember). Though, we already have some heuristic detection of that situation, so perhaps we could tweak that.

Anyhow, I think it's a good idea but it would definitely need some benchmarking. Ideally we'd also get a better dataset than the synthetic stuff in the Criterion benchmarks we have...

**Reply by CosmicHorrorDev** [reply]
Those are fair points. You do have the right idea with it essentially being speculative execution.

I do think you made a really good point that switching to something like BLAKE3 would make more sense and I did notice that there was already some code that was setting up BLAKE3 support. Would you be open to me working on finishing off that implementation instead? It does seem silly to have a special case to try and speed up MD4 hashing when switching to BLAKE3 would be a lot more straightforward

**Reply by goffrie** [reply]
Note that the existing stuff references BLAKE2 (not 3) because that's what upstream librsync supports, and has a proper magic number for. I'd be willing to accept BLAKE3 as an (opt-in) extension though.
